### PR TITLE
Update readme with workload identity based authentication for GCR and GAR

### PR DESCRIPTION
**Pull request by dineshba**
As keyless authentication is recommended, added instructions for GCR and GAR.

##### registry login using access token:
- https://cloud.google.com/container-registry/docs/advanced-authentication#token
##### workload identity:
- https://cloud.google.com/iam/docs/workload-identity-federation
- https://github.com/google-github-actions/auth#usage
- https://github.com/google-github-actions/auth#setting-up-workload-identity-federation

### CODE REVIEW

**Review by crazy-max — APPROVED** [review]
LGTM. PTAL @jonjohnsonjr

**Comment by dineshba** [comment]
@crazy-max  @jonjohnsonjr Can we merge this PR ?

**Comment by jonjohnsonjr** [comment]
Seems fine to me, @sethvargo does this look right?

File: README.md [hunk 1]
```diff
@@ -167,8 +171,48 @@ jobs:
           password: ${{ secrets.GCR_JSON_KEY }}
 ```
 
+#### Workload identity federation based authentication
```
Comment (sethvargo) [inline README.md:174]: Nit: I would prefer if WIF was _first_ since it's the preferred method

File: README.md [hunk 2]
```diff
@@ -167,8 +171,48 @@ jobs:
...
+  push:
+    branches: master
+
+jobs:
+  login:
+    runs-on: ubuntu-latest
+    steps:
+      - id: 'auth'
+      name: 'Authenticate to Google Cloud'
+      uses: 'google-github-actions/auth@v0.4.1'
```
Comment (sethvargo) [inline README.md:191]:
```suggestion
      uses: 'google-github-actions/auth@v0'
```

File: README.md [hunk 3]
```diff
@@ -167,8 +171,48 @@ jobs:
...
+on:
+  push:
+    branches: master
+
+jobs:
+  login:
+    runs-on: ubuntu-latest
+    steps:
+      - id: 'auth'
+      name: 'Authenticate to Google Cloud'
```
Comment (sethvargo) [inline README.md:190]: I think this indentation is off by two spaces for this entire section

File: README.md [hunk 4]
```diff
@@ -197,6 +241,43 @@ jobs:
 > Replace `<location>` with the regional or multi-regional [location](https://cloud.google.com/artifact-registry/docs/repo-organize#locations)
 > of the repository where the image is stored.
 
+#### Workload identity federation based authentication
```
Comment (sethvargo) [inline README.md:244]: Same about putting WIF first

File: README.md [hunk 5]
```diff
@@ -197,6 +241,43 @@ jobs:
...
+  push:
+    branches: master
+
+jobs:
+  login:
+    runs-on: ubuntu-latest
+    steps:
+      - id: 'auth'
+      name: 'Authenticate to Google Cloud'
+      uses: 'google-github-actions/auth@v0.4.1'
```
Comment (sethvargo) [inline README.md:261]:
```suggestion
      uses: 'google-github-actions/auth@v0'
```

File: README.md [hunk 6]
```diff
@@ -197,6 +241,43 @@ jobs:
...
+on:
+  push:
+    branches: master
+
+jobs:
+  login:
+    runs-on: ubuntu-latest
+    steps:
+      - id: 'auth'
+      name: 'Authenticate to Google Cloud'
```
Comment (sethvargo) [inline README.md:260]: Indentation

File: README.md [hunk 7]
```diff
@@ -197,6 +241,43 @@ jobs:
...
+        service_account: 'your-service-account@my-project.iam.gserviceaccount.com'
+      
+      - name: Login to GAR
+        uses: docker/login-action@v1
+        with:
+          registry: <location>-docker.pkg.dev
+          username: oauth2accesstoken
+          password: ${{ steps.auth.outputs.access_token }}
+```
+> Replace `<workload_identity_provider>` with configured workload identity provider
```
Comment (sethvargo) [inline README.md:274]: I don't see this defined above

File: README.md [hunk 8]
```diff
@@ -197,6 +241,43 @@ jobs:
...
+
+> Replace `<service_account>` with configured service account in workload identity provider which has access to push to GCR
```
Comment (sethvargo) [inline README.md:276]: I don't see this defined above

File: README.md [hunk 9]
```diff
@@ -167,8 +171,48 @@ jobs:
...
+#### Workload identity federation based authentication
+
+Configure the workload identity federation for github actions in gcloud (for steps, [refer here](https://github.com/google-github-actions/auth#setting-up-workload-identity-federation)). In the steps, your service account should the ability to push to GCR. Then use google-github-actions/auth action for authentication using workload identity like below:
+
+```yaml
+name: ci
+
+on:
+  push:
+    branches: master
```
Comment (sethvargo) [inline README.md:183]: Prefer `main` over `master`

File: README.md [hunk 10]
```diff
@@ -197,6 +241,43 @@ jobs:
...
+#### Workload identity federation based authentication
+
+Configure the workload identity federation for github actions in gcloud (for steps, [refer here](https://github.com/google-github-actions/auth#setting-up-workload-identity-federation)). In the steps, your service account should the ability to push to GAR. Then use google-github-actions/auth action for authentication using workload identity like below:
+
+```yaml
+name: ci
+
+on:
+  push:
+    branches: master
```
Comment (sethvargo) [inline README.md:253]: Prefer `main` over `master`

File: README.md [hunk 11]
```diff
@@ -167,8 +171,48 @@ jobs:
...
+      
+      - name: Login to GCR
+        uses: docker/login-action@v1
+        with:
+          registry: gcr.io
+          username: oauth2accesstoken
+          password: ${{ steps.auth.outputs.access_token }}
+```
+
+> Replace `<workload_identity_provider>` with configured workload identity provider
```
Comment (sethvargo) [inline README.md:205]: I don't see this defined above

File: README.md [hunk 12]
```diff
@@ -167,8 +171,48 @@ jobs:
...
+
+> Replace `<service_account>` with configured service account in workload identity provider which has access to push to GCR
```
Comment (sethvargo) [inline README.md:207]: I don't see this defined above

File: README.md (hunk 2 above)
Comment (dineshba) [reply README.md]: Fixed

File: README.md (hunk 3 above)
Comment (dineshba) [reply README.md]: Updated

File: README.md (hunk 8 above)
Comment (dineshba) [reply README.md]: Fixed

File: README.md (hunk 9 above)
Comment (dineshba) [reply README.md]: Changed in all the places

File: README.md (hunk 10 above)
Comment (dineshba) [reply README.md]: Changed in all the places

File: README.md (hunk 11 above)
Comment (dineshba) [reply README.md:243]: Fixed

File: README.md (hunk 12 above)
Comment (dineshba) [reply README.md:245]: Fixed

**Review by sethvargo — APPROVED** [review]

**Review by crazy-max — CHANGES_REQUESTED** [review]
It appears your commits messages are missing a DCO sign-off, causing the DCO check to fail.

We require all commit messages to have a Signed-off-by line with your name and e-mail, which looks something like:

```
Signed-off-by: YourFirsName YourLastName <yourname@example.org>
```

There is no need to open a new pull request, but to fix this (and make CI pass), you need to amend the commit(s) in this pull request, and "force push" the amended commit.

Unfortunately, it's not possible to do so through GitHub's web UI, so this needs to be done through the git commandline.

You can find some instructions in the output of the DCO check (which can be found in the "checks" tab on this pull request), as well as in the [Moby contributing guide](https://github.com/moby/moby/blob/v20.10.11/docs/contributing/set-up-git.md).

Steps to do so "roughly" come down to:

1. Set your name and e-mail in git's configuration:
   ```
   git config --global user.name "YourFirstName YourLastName"
   git config --global user.email "yourname@example.org"
   ```
   (Make sure to use your real name (not your GitHub username/handle) and e-mail)

2. Clone your fork locally

3. Check out the branch associated with this pull request

4. Sign-off and amend the existing commit(s)
   ```
   git commit --amend --no-edit --signoff
   ```
   If your pull request contains multiple commits, either squash the commits (if needed) or sign-off each individual commit.

5. _Force push_ your branch to GitHub (using the `--force` or [`--force-with-lease`](https://stackoverflow.com/questions/52823692/git-push-force-with-lease-vs-force) flags) to update the pull request.

Sorry for the hassle (I wish GitHub would make this a bit easier to do), and let me know if you need help or more detailed instructions!

**Comment by dineshba** [comment]
Hi @crazy-max Added the missing sign-off and squashed into one commit. Please review

**Review by crazy-max — APPROVED** [review]
LGTM thanks!

**Pull request merged**